ip_proofs/src/tipa/mod.rs

```diff
@@ -0,0 +1,499 @@
+
+use crate::{
+    gipa::{GIPAProof, GIPA},
+    Error,
+};
+use dh_commitments::{
+    afgho16::{AFGHOCommitmentG1, AFGHOCommitmentG2},
+    pedersen::PedersenCommitment,
+    DoublyHomomorphicCommitment,
+};
+use inner_products::{InnerProduct, MultiexponentiationInnerProduct};
+
+pub mod structured_scalar_message;
+
+//TODO: Could generalize: Don't need TIPA over G1 and G2, would work with G1 and G1 or over different pairing engines
+pub trait TIPACompatibleSetup {}
+
+impl<G: Group> TIPACompatibleSetup for PedersenCommitment<G> {}
+impl<P: PairingEngine> TIPACompatibleSetup for AFGHOCommitmentG1<P> {}
+impl<P: PairingEngine> TIPACompatibleSetup for AFGHOCommitmentG2<P> {}
+
+//TODO: May need to add "reverse" MultiexponentiationInnerProduct to allow for MIP with G2 messages (because TIP hard-coded G1 left and G2 right)
+pub struct TIPA<IP, LMC, RMC, IPC, P, D> {
+    _inner_product: PhantomData<IP>,
+    _left_commitment: PhantomData<LMC>,
+    _right_commitment: PhantomData<RMC>,
+    _inner_product_commitment: PhantomData<IPC>,
+    _pair: PhantomData<P>,
+    _digest: PhantomData<D>,
+}
+
+pub struct TIPAProof<IP, LMC, RMC, IPC, P, D>
+where
+    D: Digest,
+    P: PairingEngine,
+    IP: InnerProduct<
+        LeftMessage = LMC::Message,
+        RightMessage = RMC::Message,
+        Output = IPC::Message,
+    >,
+    LMC: DoublyHomomorphicCommitment + TIPACompatibleSetup,
+    RMC: DoublyHomomorphicCommitment<Scalar = LMC::Scalar> + TIPACompatibleSetup,
+    IPC: DoublyHomomorphicCommitment<Scalar = LMC::Scalar>,
+    RMC::Message: MulAssign<LMC::Scalar>,
+    IPC::Message: MulAssign<LMC::Scalar>,
+    RMC::Key: MulAssign<LMC::Scalar>,
+    IPC::Key: MulAssign<LMC::Scalar>,
+    RMC::Output: MulAssign<LMC::Scalar>,
+    IPC::Output: MulAssign<LMC::Scalar>,
+{
+    gipa_proof: GIPAProof<IP, LMC, RMC, IPC, D>,
+    final_ck: (LMC::Key, RMC::Key),
+    final_ck_proof: (P::G2Projective, P::G1Projective),
+    _pair: PhantomData<P>,
+}
+
+pub struct SRS<P: PairingEngine> {
+    g_alpha_powers: Vec<P::G1Projective>,
+    h_beta_powers: Vec<P::G2Projective>,
+    g_beta: P::G1Projective,
+    h_alpha: P::G2Projective,
+}
+
+pub struct VerifierSRS<P: PairingEngine> {
+    g: P::G1Projective,
+    h: P::G2Projective,
+    g_beta: P::G1Projective,
+    h_alpha: P::G2Projective,
+}
+
+impl<P: PairingEngine> SRS<P> {
+    pub fn get_commitment_keys(&self) -> (Vec<P::G2Projective>, Vec<P::G1Projective>) {
+        let ck_1 = self.h_beta_powers.iter().step_by(2).cloned().collect();
+        let ck_2 = self.g_alpha_powers.iter().step_by(2).cloned().collect();
+        (ck_1, ck_2)
+    }
+
+    pub fn get_verifier_key(&self) -> VerifierSRS<P> {
+        VerifierSRS {
+            g: self.g_alpha_powers[0].clone(),
+            h: self.h_beta_powers[0].clone(),
+            g_beta: self.g_beta.clone(),
+            h_alpha: self.h_alpha.clone(),
+        }
+    }
+}
+
+impl<IP, LMC, RMC, IPC, P, D> TIPA<IP, LMC, RMC, IPC, P, D>
```

Comment:
Would it make the code more modular/abstracted if you used https://github.com/scipr-lab/poly-commit/tree/master/src/kzg10 ?

(I don't know the answer, because some computations might be different)

Reply:
There are some extra computations we need. For example, we need to support shifting by a scalar `r`. That can still be made to work on top of a KZG commitment abstraction. However, the problem is that the library is fixed to commit to G1, whereas we need KZG commitments on both G1 and G2.